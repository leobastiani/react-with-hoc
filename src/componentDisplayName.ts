import { ComponentType } from "react";

export const componentDisplayName = {
  get(component: ComponentType<any>): string {
    // https://github.com/jestjs/jest/blob/3738e3f038c8a8854575f44a806ee139d2c15cdd/packages/pretty-format/src/plugins/ReactElement.ts#L36C46-L36C53
    return component.displayName || component.name || "Unknown";
  },
  set(displayName: string, component: ComponentType<any>): void {
    component.displayName = displayName;
  },
};
