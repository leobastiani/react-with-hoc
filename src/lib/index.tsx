export * from "./componentDisplayName";
export * from "./hocNameForWithStyle";
export * from "./newHoc";
