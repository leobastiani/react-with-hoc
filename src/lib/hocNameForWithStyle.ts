import { ComponentType } from "react";
import { componentDisplayName } from "./componentDisplayName";
import { GetHocArgs, HocDefinition, NewHocReturn } from "./newHoc";

export function createHocNameFunction<TNewHocReturn extends NewHocReturn<any>>(
  selector: (...args: GetHocArgs<TNewHocReturn>) => unknown | undefined
) {
  return function hocName(
    {
      Component,
      hoc,
    }: {
      Component: ComponentType<any>;
      hoc: HocDefinition<GetHocArgs<TNewHocReturn>>;
    },
    ...args: GetHocArgs<TNewHocReturn>
  ): string {
    function parseArg(arg: any): string | undefined {
      if (typeof arg === "string") {
        return arg;
      }
      if (typeof arg === "number") {
        return String(arg);
      }
      if (typeof arg === "function") {
        return arg.name || "[unnamed function]";
      }
      if (Array.isArray(arg)) {
        return arg
          .map((a) => parseArg(a))
          .filter(Boolean)
          .join(".");
      }
      if (arg === undefined || typeof arg === "function") {
        return undefined;
      }
      if (arg === null) {
        return "null";
      }
      if (typeof arg === "object") {
        return Object.keys(arg).join(".");
      }
      return String(arg);
    }

    const nameByArgs = ((): string | undefined => {
      const arg = selector(...args);

      return parseArg(arg);
    })();

    return `${[hoc.name, nameByArgs]
      .filter(Boolean)
      .join(".")}(${componentDisplayName.get(Component)})`;
  };
}

export const defaultHocName = createHocNameFunction(
  (firstArg: unknown) => firstArg
);
