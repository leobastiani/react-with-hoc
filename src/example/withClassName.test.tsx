import { render } from "@testing-library/react";
import { componentDisplayName } from "../lib/componentDisplayName";
import { withClassName } from "./withClassName";

function Example({ className }: { className: string }): JSX.Element {
  return (
    <>
      <pre>{JSON.stringify({ className })}</pre>
    </>
  );
}

it("withClassName name with no args", () => {
  const Component = withClassName()(Example);
  expect(componentDisplayName.get(Component)).toBe("withClassName(Example)");
});

it("withClassName name with string", () => {
  const Component = withClassName("someClassName")(Example);
  expect(componentDisplayName.get(Component)).toBe(
    "withClassName.someClassName(Example)"
  );
});

it("withClassName name with function", () => {
  const Component = withClassName(({ x }: { x: boolean }) =>
    x ? "someClassName" : ""
  )(Example);
  expect(componentDisplayName.get(Component)).toBe(
    "withClassName.[unnamed function](Example)"
  );
});

it("withClassName name with array of strings", () => {
  const Component = withClassName(["someClassName1", "someClassName2"])(
    Example
  );
  expect(componentDisplayName.get(Component)).toBe(
    "withClassName.someClassName1.someClassName2(Example)"
  );
});

it("withClassName name with array mixed", () => {
  const Component = withClassName([
    "someClassName1",
    ({ x }: { x: boolean }): string => (x ? "someClassName" : ""),
  ])(Example);
  expect(componentDisplayName.get(Component)).toBe(
    "withClassName.someClassName1.[unnamed function](Example)"
  );
});

describe.each([undefined, "someClass", ["someClass1", "someClass2"]])(
  "when className={%s}",
  (className) => {
    it("with no args", () => {
      const Component = withClassName()(Example);
      render(<Component className={className} />);
      expect(document.body.children[0].innerHTML).toMatchSnapshot();
    });

    it("with string", () => {
      const Component = withClassName("someOtherClass")(Example);
      render(<Component className={className} />);
      expect(document.body.children[0].innerHTML).toMatchSnapshot();
    });

    it("with function", () => {
      const Component = withClassName(({ x }: { x?: number }) =>
        x ? "someOtherClass1" : "someOtherClass2"
      )(Example);
      render(<Component className={className} />);
      expect(document.body.children[0].innerHTML).toMatchSnapshot();
    });

    it("with array of strings", () => {
      const Component = withClassName(["someOtherClass1", "someOtherClass2"])(
        Example
      );
      render(<Component className={className} />);
      expect(document.body.children[0].innerHTML).toMatchSnapshot();
    });

    it("with array mixed", () => {
      const Component = withClassName<{ x?: number }>([
        "someOtherClass1",
        ({ x }): string => (x ? "someOtherClass2" : "someOtherClass3"),
        ({ x }): string[] =>
          x
            ? ["someOtherClass4", "someOtherClass5"]
            : ["someOtherClass6", "someOtherClass7"],
      ])(Example);
      render(<Component className={className} />);
      expect(document.body.children[0].innerHTML).toMatchSnapshot();
    });
  }
);
