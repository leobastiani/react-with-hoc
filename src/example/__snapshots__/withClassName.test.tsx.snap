// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when className={[ 'someClass1', 'someClass2' ]} with array mixed 1`] = `"<pre>{"className":"someOtherClass1 someOtherClass3 someOtherClass6 someOtherClass7 someClass1 someClass2"}</pre>"`;

exports[`when className={[ 'someClass1', 'someClass2' ]} with array of strings 1`] = `"<pre>{"className":"someOtherClass1 someOtherClass2 someClass1 someClass2"}</pre>"`;

exports[`when className={[ 'someClass1', 'someClass2' ]} with function 1`] = `"<pre>{"className":"someOtherClass2 someClass1 someClass2"}</pre>"`;

exports[`when className={[ 'someClass1', 'someClass2' ]} with no args 1`] = `"<pre>{"className":"someClass1 someClass2"}</pre>"`;

exports[`when className={[ 'someClass1', 'someClass2' ]} with string 1`] = `"<pre>{"className":"someOtherClass someClass1 someClass2"}</pre>"`;

exports[`when className={someClass} with array mixed 1`] = `"<pre>{"className":"someOtherClass1 someOtherClass3 someOtherClass6 someOtherClass7 someClass"}</pre>"`;

exports[`when className={someClass} with array of strings 1`] = `"<pre>{"className":"someOtherClass1 someOtherClass2 someClass"}</pre>"`;

exports[`when className={someClass} with function 1`] = `"<pre>{"className":"someOtherClass2 someClass"}</pre>"`;

exports[`when className={someClass} with no args 1`] = `"<pre>{"className":"someClass"}</pre>"`;

exports[`when className={someClass} with string 1`] = `"<pre>{"className":"someOtherClass someClass"}</pre>"`;

exports[`when className={undefined} with array mixed 1`] = `"<pre>{"className":"someOtherClass1 someOtherClass3 someOtherClass6 someOtherClass7"}</pre>"`;

exports[`when className={undefined} with array of strings 1`] = `"<pre>{"className":"someOtherClass1 someOtherClass2"}</pre>"`;

exports[`when className={undefined} with function 1`] = `"<pre>{"className":"someOtherClass2"}</pre>"`;

exports[`when className={undefined} with no args 1`] = `"<pre>{"className":""}</pre>"`;

exports[`when className={undefined} with string 1`] = `"<pre>{"className":"someOtherClass"}</pre>"`;
